lightning/src/routing/network_graph.rs

```diff
@@ -428,6 +428,75 @@ impl Readable for NodeInfo {
 	}
 }
 
+/// Allows for updating user's network metadata.
+pub trait RouteFeePenalty {
+	/// Gets a channel's fee penalty based on its channel_id (as stored in a NetworkGraph object).
+	fn get_channel_fee_penalty(&self, chan_id: u64) -> u64;
+	/// Informs metadata object that a route has successfully executed its payment.
+	fn route_succeeded(&mut self, route: Vec<RouteHop>);
+	/// Informs metadata object that a route has failed to execute a payment.
+	fn route_failed(&mut self, route: Vec<RouteHop>, failed_hop: RouteHop);
+}
+
+/// A default metadata object that is used to implement the default functionality for the
+/// NetworkTracker trait. A user could  make their own Metadata object and extend the
+/// functionality of it by implementing other functions/traits for their metadata.
+pub struct DefaultMetadata {
+	/// A list of failed channels. Maps channel_id (as specified in the NetworkGraph object) to
+	/// the number of successful routes to participate before being removed from the list. All
+	/// channels in failed_channels are assumed to have a penalty of u64::max.
+	failed_channels: BTreeMap<u64, u64>,
+}
+
+impl DefaultMetadata {
+	/// Instantiates an instance of DefaultMetadata
+	pub fn new() -> DefaultMetadata {
+		DefaultMetadata { failed_channels: std::collections::BTreeMap::new() }
+	}
+}
+
+impl RouteFeePenalty for DefaultMetadata {
+	fn get_channel_fee_penalty(&self, chan_id: u64) -> u64 {
+		if self.failed_channels.get(&chan_id) == None {
+			return 0;
+		} else {
+			return u64::max_value();
+		}
+	}
+
+	fn route_succeeded(&mut self, route: Vec<RouteHop>) {
+		for route_hop in route {
+			let chan_id = route_hop.short_channel_id;
+			let mut can_remove = false;
+			if let Some(successes_needed) = self.failed_channels.get_mut(&chan_id) {
+				*successes_needed = *successes_needed - 1;
+				can_remove = *successes_needed == 0;
+			}
+			if can_remove {
+				self.failed_channels.remove(&chan_id);
+			}
+		}
+	}
+
+	fn route_failed(&mut self, route: Vec<RouteHop>, failed_hop: RouteHop) {
+		for route_hop in route {
+			if route_hop == failed_hop {
+				*self.failed_channels.entry(failed_hop.short_channel_id).or_insert(5) += 1;
```

Comment:
TBH this feels a bit weird and arbitrary - you need 5 successes to un-fail a channel but once we have even 1 failure we refuse to route through a channel at all. I think maybe a more interesting approach would be to consider the time of the last failure and de-prioritize exponentially less as a failure ages, removing it entirely if we succeed. Because we cannot call current-time functions in the core library, we can't do that exactly, however something that may approximate would be to fail a channel until the next success - ie if we fail a channel, we wont use it the next time we route, but we'll use it any time after that. A user could, of course, do something better themselves by re-implementing the API. Separately, we could maybe add a feature flag for "we can call gettimeofday()".

Reply:
Yes, I agree that it feels weird/arbitrary. I like the idea of having the exponential backoff too, but as mentioned we do not have access to time. For now, I am implementing "the opposite" where on success routing, all channels are removed from the failed_channels list, and on failure, channels have a "consecutive fails since last successful payment" count incremented.

Reply:
Right, but because we fully-refuse to route through a channel after its failed once, we should never be able to send through it ever again if it fails once (because we'll never succeed routing through it if we never try...).